contracts/registries/YRegistry.sol

```diff
@@ -0,0 +1,213 @@
+    // (current ones can be obtained via getVaults + getVaultInfo)
+  }
+
+  function setVault(address _vault) internal {
+    require(_vault.isContract(), "Vault is not a contract");
+    // Checks if vault is already on the array
+    require(!vaults.contains(_vault), "Vault already exists");
+    // Adds unique _vault to vaults array
+    vaults.add(_vault);
+  }
+
+  function setWrappedVault(address _vault, address _wrappedVault) internal {
+    require(_wrappedVault.isContract(), "WrappedVault is not a contract");
+    wrappedVaults[_vault] = _wrappedVault;
+  }
+
+  function setDelegatedVault(address _vault) internal {
+    // TODO Is there any way to check if a vault is delegated
+    isDelegatedVault[_vault] = true;
+  }
+
+  function setController(address _controller) internal {
+    // Adds Controller to controllers array
+    if (!controllers.contains(_controller)) {
+      controllers.add(_controller);
+    }
+  }
+
+  function getVaultData(address _vault) internal view returns (
+    address controller,
+    address token,
+    address strategy,
+    bool isWrapped,
+    bool isDelegated
+  ) {
+    address vault = _vault;
+    isWrapped = wrappedVaults[_vault] != address(0);
+    if (isWrapped) {
+      vault = wrappedVaults[_vault];
+    }
+    isDelegated = isDelegatedVault[vault];
+    
+    // Get values from controller
+    controller = IVault(vault).controller();
+    if (isWrapped && IVault(vault).underlying() != address(0)) {
+      token = IVault(_vault).token(); // Use non-wrapped vault
+    } else {
+      token = IVault(vault).token();
+    }
+    
+    if (isDelegated) {
+      strategy = IController(controller).strategies(vault);
+    } else {
+      strategy = IController(controller).strategies(token);
+    }
+    
+    // Check if vault is set on controller for token
+    address controllerVault = address(0);
+    if (isDelegated) {
+      controllerVault = IController(controller).vaults(strategy);
+    } else {
+      controllerVault = IController(controller).vaults(token);
+    }
+    require(controllerVault == vault, "Controller vault address does not match"); // Might happen on Proxy Vaults
+    
+    // Check if strategy has the same token as vault
+    if (isWrapped) {
+      address underlying = IVault(vault).underlying();
+      require(underlying == token, "WrappedVault token address does not match"); // Might happen?
+    } else if (isDelegated) {
+
+    } else {
+      address strategyToken = IStrategy(strategy).want();
+      require(token == strategyToken, "Strategy token address does not match"); // Might happen?
+    }
+
+    return (controller, token, strategy, isWrapped, isDelegated);
+  }
+
+  // Vaults getters
+  function getVault(uint index) external view returns (address vault) {
+    return vaults.at(index);
+  }
+
+  function getVaultsLength() external view returns (uint) {
+    return vaults.length();
+  }
+
+  function getVaults() external view returns (address[] memory) {
```

Comment:
it would be great to have this method return an array of structs of vaultData for all vaults, thinking of UIs that want to call a single method to get all vaults data in one shot.  With this design it needs two calls to do something similar like https://feel-the-yearn.app/
first do getVaults, then getVaultData() for each vault in parallel per address, maybe a batch method could be helpful?

Reply:
Coding this would probably bloat the bytecode significantly, making it more costly to deploy. The front end code could also probably take advantage of this too:
```javascript
const vaults = Promise.map(await contract.getVaults(), (vault_address) => await contract.getVaultData(vault_address));
```

Reply:
yeah, its a tradeoff for deploying vrs simplicity on UI, im fine with removing it if we want to make deployment less expensive

Reply:
Not sure how much more expensive it would be, but I'm in favor of adding simplicity to the UI. 
Since it will require less calls to `infura, etc` and it's a HUGE mess if any of the `await contract.getVaultData(vault_address)` fails :/

Reply:
Instead of promises, you can bundle several `eth_call` into one using [multicall](https://github.com/makerdao/multicall).